### PR TITLE
Effect selector widget

### PR DESCRIPTION
**Pull request by Be-ing**
Add a new QComboBox widget to select an effect. This has several advantages:
* Next/previous effect and eject effect buttons can be removed from skins, freeing valuable space
* If the user wants to load a specific effect, they do not need to cycle through effects to do so.
* The list of available effects is visible to the user.
* The list of available effects is in predictable alphabetical order.

It is easy to replace EffectName elements in skin XML with the new EffectSelector element. All that is required is changing the element name. The skin will need to define styles for WEffectSelector to make the colors match the skin instead of the system Qt colors.

### CODE REVIEW

**Comment by Be-ing** [comment]
I am thinking we should remove the effect full name versus short name distinction. I think we only need short names and long descriptions.

**Comment by daschuer** [comment]
> I am thinking we should remove the effect full name versus short name distinction. I think we only need short names and long descriptions.

The issue now is that we have no rule, what is short and what is long. 
I can imagine when that a skin can make use of the short names, when labeling knobs and the space is relay limited. In this case the Lables can be abbreviated up to just before unrecognizable.  
On the other hand a well recognizable name is very useful if some the space is there, like the Shade Effect selector. 

Can we come to a rule, that is usable for Translators and considered proportional fonts? 
// Trans late not longer than "XXXXXXX"

File: src/effects/effectsmanager.cpp
Comment (daschuer) [inline src/effects/effectsmanager.cpp]: Using QPair is an antipattern. Please create your own proper named struct or class.

File: src/effects/effectsmanager.cpp
Comment (daschuer) [inline src/effects/effectsmanager.cpp]: Use QString::localeAwareCompare()

File: src/effects/effectsmanager.cpp
Comment (daschuer) [inline src/effects/effectsmanager.cpp]: Oy! QPair ;-)

File: src/effects/effectsmanager.cpp
Comment (daschuer) [inline src/effects/effectsmanager.cpp]: This can become a range based loop

Comment (daschuer) [inline src/effects/effectsmanager.cpp]: Oh no, I see.

File: src/effects/effectsmanager.cpp
Comment (daschuer) [inline src/effects/effectsmanager.cpp]: I think you can move out the common inner part of the Prev and Next functions.

**Comment by daschuer** [comment]
I think the effect selector is a candidate for the long names by default. 
If someone has no space, an option for the short names would be nice.  

Is it possible to show the short names in the Box and the long names and short in the selector? 
"Short (Long Name)" 

A tool tip showing the description might be also usable.

**Comment by Be-ing** [comment]
> I can imagine when that a skin can make use of the short names, when labeling knobs and the space is relay limited. In this case the Lables can be abbreviated up to just before unrecognizable.

You are conflating effect names and parameter names. I think short names for parameters should stay, but we should move to one short name for effects.

> Is it possible to show the short names in the Box and the long names and short in the selector?
"Short (Long Name)"

That was what I intended to do initially. However, I realized it would be a bit confusing to see the text change when selecting a new effect versus seeing the selected effect name. "Short (Long Name)" would be overcomplicated and unnecessarily long. The default behavior of QComboBox is to expand to the width of the longest item in the list. I think this behavior makes sense and should be kept, so the list items should be kept short.

> A tool tip showing the description might be also usable.

That is already implemented in this PR, which is why I think long names are unnecessary.

**Comment by Be-ing** [comment]
> You are conflating effect names and parameter names. I think short names for parameters should stay, but we should move to one short name for effects.

Actually, the same applies to effect parameters. Space will always be limited where they are displayed, so I don't think there is ever a need for an extended name. The description is sufficient.

File: src/effects/effectsmanager.cpp
Comment (timrae) [inline src/effects/effectsmanager.cpp]: Why?? :-/

Comment (timrae) [inline src/effects/effectsmanager.cpp]: Even as a typedef: `using NameIdPair = QPair<QString, QString>`

Comment (Be-ing) [inline src/effects/effectsmanager.cpp]: I think I'll refactor this so EffectManager keeps a QList<EffectManifest> sorted alphabetically. But yeah, I don't understand what the issue with QPair is.

Comment (daschuer) [inline src/effects/effectsmanager.cpp]: QPair is bad, because you cannot see what .first and .second is.
It requires explanation on every usage. 
class NameIdPair {
  public: 
    QString Id;
    QString Name; 
}

would be self explainig

File: src/widget/weffectselector.cpp
Comment (rryan) [inline src/widget/weffectselector.cpp]: Instead of having a null string constant, you can use the fact the QVariant and QString both have ```isNull``` methods for when their no-args constructor is used. 

```QString("").isNull() == false``` and ```QString().isNull() == true```

You can use ```QVariant()``` and then check ```QVariant::isNull``` instead of

File: src/widget/weffectselector.cpp
Comment (rryan) [inline src/widget/weffectselector.cpp]: check for null m_pChainSlot early and exit?

File: src/widget/weffectselector.cpp
Comment (rryan) [inline src/widget/weffectselector.cpp]: I think a widget shouldn't have the instantiate/replace logic in it -- could you add a ```EffectRack::loadEffect(chainslotnumber, effectslotnumber, effectid)``` (alongside loadNextEffect/loadPrevEffect) method, then call it here?

File: src/widget/weffectselector.cpp
Comment (rryan) [inline src/widget/weffectselector.cpp]: It turns out this only applies to user-entered data, not programmatically entered data -- so you can remove it.

**Comment by daschuer** [comment]
Now (the same in master) the Filter effects are started silence, because the super knob is adopted to avoid going out of sync. 
I am not sure if I can get used to it.  I think I like the old behaviour more that every effect starts with its reasonable default parameters and the super knob has to snap in by passing the defaults. 
This might be a problem when there is no super knob visible. for example if you load the Effect to the EQ knobs. 

I also have mixed feelings if we move the Super knob to the end. This will move out the most important parameter from the parameter knobs if you have only three parameter knobs. 

We can also not guarantee this assumption if we consider external effects. 

On the other hand I fully understand your requirement to not have a redundant knob in single parameter mode.  

What alternatives we have? Do your really need the super knob in single parameter mode? Would it be possible to use it as 4th parameter in that case?

**Comment by daschuer** [comment]
Ups, I commented the wrong PR, moving my comment.

File: src/effects/effectsmanager.cpp
Comment (timrae) [inline src/effects/effectsmanager.cpp]: In this case you can just use a typedef though...

Comment (daschuer) [inline src/effects/effectsmanager.cpp]: You cannot get rid of the meaningless `first` and `second` members.

File: src/effects/effectsmanager.cpp
Comment (Be-ing) [inline src/effects/effectsmanager.cpp]: That's so little code that I don't think it would be worth it.

File: src/effects/effectsmanager.cpp
Comment (timrae) [inline src/effects/effectsmanager.cpp]: True, I agree.

**Comment by Be-ing** [comment]
Now that I have gone through the trouble of removing the QPairs, I understand why it is not a good idea to use them.

**Review by rryan — CHANGES_REQUESTED** [review]
Looks great! LGTM other than a couple nits.

**Review by rryan — APPROVED** [review]

**Comment by rryan** [comment]
@daschuer I think all of your concerns were addressed (?) so I'll merge this once CI is green.

**Comment by daschuer** [comment]
@Be-ing: 
If you like, you could brake long lines, some are leading to vertical scrolling even in GitHub? 

Am manual test can't hurt. Do you have a testing skin? 
But I do not want to delay this ... LGTM!
Thank you for the work.

**Comment by Be-ing** [comment]
I'll update #1063 after this is merged.

**Comment by daschuer** [comment]
I have just tested it and it works nice.

 I have noticed three small issues: 
* There is no elide when the Effect name is longer than the box. In th other effect selector we use a middle elide which works good. 
* The effect description is displayed as a single line tooltip, which continues outside small resolution displays. It would be nice to break them at a reasonable width.
* The developer tooltips are missing.

**Comment by Be-ing** [comment]
> There is no elide when the Effect name is longer than the box. In th other effect selector we use a middle elide which works good.

Is this because there are not translated shortNames yet? shortNames should not need eliding.

I'll look into the other issues.

**Comment by daschuer** [comment]
I think the box should elide in any case. This will be required for LV2 effects.

**Comment by Be-ing** [comment]
Okay. When LV2 support is introduced we may consider adding shortNames for them.

Middle eliding seems kinda odd. Shouldn't it elide at the end?

**Comment by Be-ing** [comment]
Without eliding, the width of the widget automatically expands to the width of the widest item, at least in the context of replacing the `<Effect>` in #1063 with `<EffectSelector>`. I think this is appropriate behavior. It is then up to translators to make sure displayed names are not ridiculously long.

**Comment by daschuer** [comment]
We cannot add names for LV2 effects, because we do not know them before. 
The user could do it, but we should not force him. 
The middle elide it the Best solution for "Fancy Effect 1" and "Fancy Effect 2" which is the case for some LV2 effects:  

This is my output of lv2ls -n: 

Calf Analyzer
Calf Bass Enhancer
Calf Compensation Delay Line
Calf Compressor
Calf Deesser
Calf Emphasis
Calf Equalizer 12 Band
Calf Equalizer 5 Band
Calf Equalizer 8 Band
Calf Exciter
Calf Filter
Calf Filterclavier
Calf Flanger
Calf Fluidsynth
Calf Gate
Calf Limiter
Calf Mono Compressor
Calf Mono Input
Calf Monosynth
Calf Multi Chorus
Calf Multiband Compressor
Calf Multiband Gate
Calf Multiband Limiter
Calf Organ
Calf Phaser
Calf Pulsator
Calf Reverb
Calf Rotary Speaker
Calf Saturator
Calf Sidechain Compressor
Calf Sidechain Gate
Calf Stereo Tools
Calf Tape Simulator
Calf Transient Designer
Calf Vintage Delay
Calf Wavetable
Calf X-Over 2 Band
Calf X-Over 3 Band
Calf X-Over 4 Band
Invada Compressor (mono)
Invada Compressor (stereo)
Invada Delay Munge (mono in)
Invada Delay Munge (sum L+R in)
Invada Early Reflection Reverb (mono in)
Invada Early Reflection Reverb (sum L+R in)
Invada High Pass Filter (mono)
Invada High Pass Filter (stereo)
Invada Low Pass Filter (mono)
Invada Low Pass Filter (stereo)
Invada Input Module
Invada Meter
Invada Stereo Phaser (mono in)
Invada Stereo Phaser (stereo in)
Invada Stereo Phaser (sum L+R in)
Invada Test Tones
Invada Tube Distortion (mono)
Invada Tube Distortion (stereo)
Einfacher Verstärker
Example Metronome
Example MIDI Gate
Example Sampler
Example Scope (Mono)
Example Scope (Stereo)
A-Law Compressor
Aliasing
Allpass delay line, cubic spline interpolation
Allpass delay line, linear interpolation
Allpass delay line, noninterpolating
AM pitchshifter
Simple amplifier
Analogue Oscillator
Artificial latency
Auto phaser
Glame Bandpass Analog Filter
Glame Bandpass Filter
Bode frequency shifter
Bode frequency shifter (CV)
GLAME Butterworth Highpass
GLAME Butterworth Lowpass
Glame Butterworth X-over Filter
Chebyshev distortion
Comb Filter
Comb Splitter
Comb delay line, cubic spline interpolation
Comb delay line, linear interpolation
Comb delay line, noninterpolating
Constant Signal Generator
Crossover distortion
DC Offset Remover
Exponential signal decay
Decimator
Declipper
Simple delay line, cubic spline interpolation
Simple delay line, linear interpolation
Simple delay line, noninterpolating
Delayorama
Diode Processor
Audio Divider (Suboctave Generator)
DJ flanger
DJ EQ
DJ EQ (mono)
Dyson compressor
Fractionally Addressed Delay Line
Fast Lookahead limiter
Flanger
FM Oscillator
Foldover distortion
4 x 4 pole allpass
Fast overdrive
Frequency tracker
Gate
Giant flange
Gong model
Gong beater
GVerb
Hard Limiter
Harmonic generator
Hermes Filter
Glame Highpass Filter
Hilbert transformer
Non-bandlimited single-sample impulses
Inverter
Karaoke
L/C/R Delay
LFO Phaser
Lookahead limiter
Lookahead limiter (fixed latency)
Glame Lowpass Filter
LS Filter
Matrix: MS to Stereo
Matrix Spatialiser
Matrix: Stereo to MS
Multiband EQ
Modulatable delay
Multivoice Chorus
Higher Quality Pitch Scaler
Plate reverb
Pointer cast distortion
Rate shifter
Retro Flanger
Reverse Delay (5s max)
Ringmod with LFO
Ringmod with two inputs
Barry's Satan Maximiser
SC1
SC2
SC3
SC4
SE4
Wave shaper
Signal sifter
Sine + cosine oscillator
Single band parametric
Sinus wavewrapper
Smooth Decimator
Mono to Stereo splitter
Surround matrix encoder
State Variable Filter
Tape Delay Simulation
Transient mangler
Triple band parametric with shelves
μ-Law Compressor
Valve saturation
Valve rectifier
VyNil (Vinyl Effect)
Wave Terrain Oscillator
Crossfade
Crossfade (4 outs)
z-1
Kn0ck0ut Vocal Remover

**Comment by Be-ing** [comment]
Hm, the widget shows the English shortNames for effects that have them when I run Mixxx with fr_FR locale. Does every effect need a shortName declared in its manifest to let translators provide shortNames even if a shortName is not needed for that particular effect's English name?

File: src/widget/weffectselector.cpp
Comment (Be-ing) [inline src/widget/weffectselector.cpp]: This wraps at a rather narrow width, but I don't think there's a way to set Qt to wrap at a wider width. It's a little awkward, but it's better than having very wide tooltips.

**Comment by daschuer** [comment]
> Hm, the widget shows the English shortNames for effects that have them when I run Mixxx with fr_FR locale. Does every effect need a shortName declared in its manifest to let translators provide shortNames even if a shortName is not needed for that particular effect's English name?

Good point! I would say yes.

Did you consider to add a comment that gives a hint what short and long is?  How about the proposed "XXXXX" measure?

**Comment by Be-ing** [comment]
> Did you consider to add a comment that gives a hint what short and long is?

I'm not sure what hint we'd give. What length it would actually cut off depends on the font face, font size, and width allowed for the WEffectSelector by the skin.

> How about the proposed "XXXXX" measure?

What proposal are you referring to?

File: src/widget/weffectselector.cpp
Comment (Be-ing) [inline src/widget/weffectselector.cpp]: I tried using `<div style=\"width: 450px;\"> + description + </div>` but that did not change the wrapping width.

**Comment by Be-ing** [comment]
If every effect will have a shortName, I think the new EffectManifest::displayName method can be removed and shortName can be used in its place.

**Comment by daschuer** [comment]
Something like this: 
//: Limit the length of your translation to the rendered pixel width of "XXXXXXX"

"XXXXXXX"
"Fancy Effect"  -> too long 
"Fancy Ef." -> fit even if there are more charater. 
"花式效果" -> Fit 
"花式特殊效果" does not fit even if there are less character.

**Comment by rryan** [comment]
Let's continue to work on the effect name lengths and translation questions -- but let's not hold up this PR.

**Comment by Be-ing** [comment]
Not off the top of my head, but take a look at the [QComboBox styling documentation](http://doc.qt.io/qt-4.8/stylesheet-examples.html#customizing-qcombobox).

**Pull request merged**
